src/lxd.py

```diff
@@ -0,0 +1,321 @@
+            destination,
+        ]
+
+        try:
+            execute_command(lxc_cmd)
+        except CalledProcessError as err:
+            raise LxdError(
+                f"Unable to pull file {str(source)} from LXD instance {self.instance.name}"
+            ) from err
+
+    def get(self, filepath: str) -> str:
+        """Get content of a file in the LXD instance.
+
+        Args:
+            filepath: Path of the file in the LXD instance.
+
+        Raises:
+            LxdException: Unable to load the file from the LXD instance.
+
+        Returns:
+            The content of the file.
+        """
+        with tempfile.NamedTemporaryFile() as file:
+            self.pull(filepath, file.name)
+
+            return file.read().decode("utf-8")
+
+
+class LxdInstance:
+    """A LXD instance.
+
+    Attrs:
+        name (str): Name of LXD instance.
+        files (LxdInstanceFiles): Manager for the files on the LXD instance.
+        status (str): Status of the LXD instance.
+    """
+
+    def __init__(self, name: str, pylxd_instance: pylxd.models.Instance):
+        """Construct the LXD instance representation.
+
+        Args:
+            pylxd_instance: Instance of pylxd.models.Instance for the LXD instance.
+        """
+        self.name = name
+        self._pylxd_instance = pylxd_instance
+        self.files = LxdInstanceFiles(self._pylxd_instance)
+
+    @property
+    def status(self) -> str:
+        """Status of the LXD instance.
+
+        Returns:
+            Status of the LXD instance.
+        """
+        return self._pylxd_instance.status
+
+    def start(self, timeout: int = 30, force: bool = True, wait: bool = False) -> None:
+        """Start the LXD instance.
+
+        Args:
+            timeout: Timeout for starting the LXD instance.
+            force: Whether to force start the LXD instance.
+            wait: Whether to wait until the LXD instance started before returning.
+
+        Raises:
+            LxdException: Unable to start the LXD instance.
+        """
+        try:
+            self._pylxd_instance.start(timeout, force, wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            raise LxdError(f"Unable to start LXD instance {self.name}") from err
+
+    def stop(self, timeout: int = 30, force: bool = True, wait: bool = False) -> None:
+        """Stop the LXD instance.
+
+        Args:
+            timeout: Timeout for stopping the LXD instance.
+            force: Whether to force stop the LXD instance.
+            wait: Whether to wait until the LXD instance stopped before returning.
+
+        Raises:
+            LxdException: Unable to stop the LXD instance.
+        """
+        try:
+            self._pylxd_instance.stop(timeout, force, wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            raise LxdError(f"Unable to stop LXD instance {self.name}") from err
+
+    def execute(self, cmd: list[str], cwd: Optional[str] = None) -> Tuple[int, IO, IO]:
```

Comment:
SOmething to think about: consider whether there should be a higher level abstraction that is the actual commands that should be executed rather than just re-exposing the execute interface. E.g., think about what the charm would need to execute on the instance and make those Python functions which are then converted to commands and the output parsed etc. That way, the messy details of writing commands is abstracted from most of the code. What do you think?